src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,171 @@
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNodeAtIndex(index);
+        T oldValue = current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNodeAtIndex(index);
+        if (current.prev != null) {
+            current.prev.next = current.next;
+        } else {
+            head = current.next;
+        }
+        if (current.next != null) {
+            current.next.prev = current.prev;
+        } else {
+            tail = current.prev;
+        }
+        return current.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if (object == null ? current.value == null :
+                    object.equals(current.value)) {
+                removeNode(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        int count = 0;
+        Node<T> current = head;
+        while (current != null) {
+            count++;
+            current = current.next;
+        }
+        return count;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return head == null;
+    }
+
+    private Node<T> createNewNode(T value) {
+        return new Node<>(null, value, null);
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size()) {
+            throw new IndexOutOfBoundsException("Index "
+                    + index + " is out of bounds. Size: " + size());
+        }
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size()) {
+            throw new IndexOutOfBoundsException("Index "
+                    + index + " is out of bounds. Size: " + size());
+        }
+    }
+
+    private void addFirst(Node<T> newNode) {
+        newNode.next = head;
+        head.prev = newNode;
+        head = newNode;
+    }
+
+    private void insertAfter(Node<T> current, Node<T> newNode) {
+        newNode.prev = current;
+        newNode.next = current.next;
+        if (current.next != null) {
+            newNode.next.prev = newNode;
+        }
+        current.next = newNode;
+    }
+
+    private void removeNode(Node<T> node) {
+        if (node.prev != null) {
+            node.prev.next = node.next;
+        } else {
+            head = node.next;
+        }
+        if (node.next != null) {
+            node.next.prev = node.prev;
+        } else {
+            tail = node.prev;
+        }
+    }
+
+    private Node<T> getNodeAtIndex(int index) {
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current;
+    }
```

Comment:
checklist item
https://github.com/mate-academy/jv-linked-list/blob/master/checklist.md#dont-iterate-your-nodes-from-the-head-to-the-tail-or-vice-versa